File: app/lib/utils/password_policy.dart
/// Matches backend [STRONG_PASSWORD_MESSAGE] for admin + password reset flows.
class PasswordPolicy {
  PasswordPolicy._();

  static const String strongPasswordMessage =
      'Password must be at least 12 characters and include uppercase, lowercase, a number, and a symbol.';

  static final RegExp _special = RegExp(r'''[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?]''');

  /// Returns null if valid, otherwise the user-facing error message.
  static String? validateStrongPassword(String raw) {
    final p = raw.trim();
    if (p.length < 12) return strongPasswordMessage;
    if (!RegExp(r'[A-Z]').hasMatch(p)) return strongPasswordMessage;
    if (!RegExp(r'[a-z]').hasMatch(p)) return strongPasswordMessage;
    if (!RegExp(r'[0-9]').hasMatch(p)) return strongPasswordMessage;
    if (!_special.hasMatch(p)) return strongPasswordMessage;
    return null;
  }
}

File: backend/src/utils/password_policy.ts
/**
 * Strong password rules for admin accounts and password-reset flows.
 * Kept in one place so API routes and services stay consistent.
 */

export const STRONG_PASSWORD_MESSAGE =
  'Password must be at least 12 characters and include uppercase, lowercase, a number, and a symbol.';

const SPECIAL_RE = /[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?]/;

/**
 * Validates password strength; throws Error with STRONG_PASSWORD_MESSAGE if invalid.
 */
export const assertStrongPassword = (raw: string): void => {
  const password = (raw ?? '').trim();
  if (password.length < 12) {
    throw new Error(STRONG_PASSWORD_MESSAGE);
  }
  if (!/[A-Z]/.test(password)) {
    throw new Error(STRONG_PASSWORD_MESSAGE);
  }
  if (!/[a-z]/.test(password)) {
    throw new Error(STRONG_PASSWORD_MESSAGE);
  }
  if (!/[0-9]/.test(password)) {
    throw new Error(STRONG_PASSWORD_MESSAGE);
  }
  if (!SPECIAL_RE.test(password)) {
    throw new Error(STRONG_PASSWORD_MESSAGE);
  }
};
